src/psyclone/domain/lfric/metadata_to_arguments_rules.py

```diff
@@ -0,0 +1,614 @@
+                    cls._indirection_dofmap(
+                        function_space, cma_ops[0])
+
+            # Provide any optional arguments. These arguments are
+            # associated with the keyword arguments (basis function
+            # and differential basis function) for a function space.
+            meta_funcs = metadata.meta_funcs \
+                if metadata.meta_funcs else []
+            if [func for func in meta_funcs if func.basis_function and
+                    func.function_space == function_space]:
+                cls._basis(function_space)
+            if [func for func in meta_funcs if func.diff_basis_function
+                    and func.function_space == function_space]:
+                cls._diff_basis(function_space)
+
+        # The boundary condition kernel (enforce_bc_kernel) is a
+        # special case.
+        if metadata.procedure_name and \
+                metadata.procedure_name.lower() == "enforce_bc_code":
+            cls._field_bcs_kernel()
+
+        # The operator boundary condition kernel
+        # (enforce_operator_bc_kernel) is a special case.
+        if metadata.procedure_name and \
+                metadata.procedure_name.lower() == "enforce_operator_bc_code":
+            cls._operator_bcs_kernel()
+
+        # Reference-element properties
+        if metadata.meta_ref_element:
+            cls._ref_element_properties(metadata.meta_ref_element)
+
+        # Mesh properties
+        if metadata.meta_mesh:
+            cls._mesh_properties(metadata.meta_mesh)
+
+        # Quadrature arguments are required if one or more basis or
+        # differential basis functions are used by the kernel and a
+        # quadrature shape is supplied.
+        if metadata.meta_funcs and metadata.shapes and \
+           [shape for shape in metadata.shapes if shape in
+                const.VALID_QUADRATURE_SHAPES]:
+            cls._quad_rule(metadata.shapes)
+
+    @staticmethod
+    def _field_meta_args_on_fs(arg_types, function_space, metadata):
+        '''Utility function to return any field (plus field vector, intergrid
+        or intergrid vector) meta_args in metadata that have the same
+        type as those specified in arg_types and are on the function
+        space specified in function_space.
+
+        :param arg_types: meta_arg classes indicating which meta_arg \
+            arguments to check.
+        :type arg_types: subclass of \
+            :py:class:`psyclone.domain.lfric.kernel.CommonMetaArgMetadata` or \
+            List[subclass of \
+            :py:class:`psyclone.domain.lfric.kernel.CommonMetaArgMetadata`]
+        :param str function_space: the specified function space.
+        :param metadata: the kernel metadata.
+        :type metadata: \
+            :py:class:`psyclone.domain.lfric.kernel.LFRicKernelMetadata`
+
+        :returns: a list of meta_args.
+        :type arg_types: List[subclass of \
```

Comment:
s/subclass of//

Reply:
I like adding subclass as I think it explains what is expected. However, I have removed them all.